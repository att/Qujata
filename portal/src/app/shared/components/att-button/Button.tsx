import { useCallback } from 'react';
import cn from 'classnames';
import styles from './Button.module.scss';
import { ButtonActionType, ButtonSize, ButtonStyleType } from './Button.model';
import { ButtonCssClassBySize, ButtonCssClassByStyleType } from './Button.const';

export interface ButtonProps {
  actionType: ButtonActionType;
  ariaLabel?: string;
  className?: string;
  disabled?: boolean;
  size: ButtonSize;
  styleType: ButtonStyleType;
  onButtonClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
  children: React.ReactNode;
  form?: string;
  title?: string;
  e2eId?: string;

}

export const Button: React.FC<ButtonProps> = (props: ButtonProps) => {
  const { onButtonClick } = props;
  const onClick: (e: React.MouseEvent<HTMLButtonElement>) => void = useCallback((e: React.MouseEvent<HTMLButtonElement>): void => {
    e.currentTarget.blur();
    onButtonClick(e);
  }, [onButtonClick]);

  const ButtonElement: JSX.Element = (
    <button
      className={cn(props.className, styles.button, ButtonCssClassByStyleType.get(props.styleType), ButtonCssClassBySize.get(props.size))}
      disabled={props.disabled}
      type={props.actionType}
      aria-label={props.ariaLabel}
      onClick={onClick}
      form={props.form}
      title={props.title}
      data-cy={props.e2eId}
    >
      {props.children}
    </button>
  );

  return ButtonElement;
};
