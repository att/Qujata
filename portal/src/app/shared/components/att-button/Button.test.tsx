import { act, fireEvent, render, RenderResult } from '@testing-library/react';
import { Button, ButtonProps } from './Button';
import { ButtonActionType, ButtonSize, ButtonStyleType } from './Button.model';

describe('Button', () => {
  test('renders large primary button', () => {
    const props: ButtonProps = {
      styleType: ButtonStyleType.PRIMARY,
      actionType: ButtonActionType.BUTTON,
      size: ButtonSize.LARGE,
      onButtonClick: jest.fn(),
      children: undefined,
    };
    const { container }: RenderResult = render(<Button {...props}>large primary button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders medium primary button', () => {
    const props: ButtonProps = {
      styleType: ButtonStyleType.PRIMARY,
      actionType: ButtonActionType.BUTTON,
      size: ButtonSize.MEDIUM,
      onButtonClick: jest.fn(),
      children: undefined,
    };
    const { container }: RenderResult = render(<Button {...props}><span>medium primary button</span></Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders small primary button', () => {
    const props: ButtonProps = {
      styleType: ButtonStyleType.PRIMARY,
      actionType: ButtonActionType.BUTTON,
      size: ButtonSize.SMALL,
      onButtonClick: jest.fn(),
      children: undefined,
    };
    const { container }: RenderResult = render(<Button {...props}><span>small primary button</span></Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders small secondary button', () => {
    const props: ButtonProps = {
      styleType: ButtonStyleType.SECONDARY,
      actionType: ButtonActionType.BUTTON,
      size: ButtonSize.SMALL,
      onButtonClick: jest.fn(),
      children: undefined,
    };
    const { container }: RenderResult = render(<Button {...props}>small secondary button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders small text button', () => {
    const props: ButtonProps = {
      styleType: ButtonStyleType.TEXT,
      actionType: ButtonActionType.BUTTON,
      size: ButtonSize.NONE,
      onButtonClick: jest.fn(),
      children: undefined,
    };
    const { container }: RenderResult = render(<Button {...props}>small text button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders wrapper button', () => {
    const props: ButtonProps = {
      styleType: ButtonStyleType.WRAPPER,
      actionType: ButtonActionType.SUBMIT,
      size: ButtonSize.NONE,
      onButtonClick: jest.fn(),
      children: undefined,
    };
    const { container }: RenderResult = render(<Button {...props}>small wrapper button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('button click behavior', () => {
    const props: ButtonProps = {
      styleType: ButtonStyleType.WRAPPER,
      actionType: ButtonActionType.SUBMIT,
      size: ButtonSize.NONE,
      onButtonClick: jest.fn(),
      children: undefined,
    };
    const { container }: RenderResult = render(<Button {...props}>small wrapper button</Button>);
    act(() => {
      fireEvent.click(container.firstChild as HTMLButtonElement);
    });
    expect(props.onButtonClick).toHaveBeenCalled();
  });
});
