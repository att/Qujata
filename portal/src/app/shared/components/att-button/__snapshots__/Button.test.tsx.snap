// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders large primary button 1`] = `
<button
  class="button primary_button lg_button"
  type="button"
>
  large primary button
</button>
`;

exports[`Button renders medium primary button 1`] = `
<button
  class="button primary_button md_button"
  type="button"
>
  <span>
    medium primary button
  </span>
</button>
`;

exports[`Button renders small primary button 1`] = `
<button
  class="button primary_button sm_button"
  type="button"
>
  <span>
    small primary button
  </span>
</button>
`;

exports[`Button renders small secondary button 1`] = `
<button
  class="button secondary_button sm_button"
  type="button"
>
  small secondary button
</button>
`;

exports[`Button renders small text button 1`] = `
<button
  class="button txt_button"
  type="button"
>
  small text button
</button>
`;

exports[`Button renders wrapper button 1`] = `
<button
  class="button wrapper_button"
  type="submit"
>
  small wrapper button
</button>
`;
